### PR TITLE
Support exclude files

### PR DESCRIPTION
**Pull request by i-am-the-slime**
Adds support for exclude files which enables colocation (#602)

`excludeFiles` is an array of globs. The corresponding expanded paths will be subtracted from all included paths (implicit and explicit via `files`).

That way it's possible to keep related files close together but not ship them in the package 🥳.

Right now this depends on a forked version of spaghetto which depends on this fork. I'm not sure in what order this could be merged.

### CODE REVIEW

**Comment by f-f** [comment]
Thanks for putting this together @i-am-the-slime! I am on holiday away from screens this week, so I'll be able to review all of this once I'm back on Sunday 🙂

**Comment by f-f** [comment]
@i-am-the-slime the dependency order is `spago-core -> registry -> spago`, so we'll want to merge a PR to `spago-core` first, then one to `registry` (this one), then one to `spago`

**Comment by thomashoneyman** [comment]
I think we should use the same names (include / exclude), which would mean changing the Manifest type (and the associated dhall spec and description in SPEC.md). If you’d like to do that, great! If you’d rather I do it since I know where all the relevant files are then I can — just let me know.

**Comment by i-am-the-slime** [comment]
@thomashoneyman I won't say no to this offer. Especially because I'm having trouble with the new spago in IntelliJ which is what I'm used to using these days.

**Comment by thomashoneyman** [comment]
I ended up going with `includeFiles` and `excludeFiles` — once I was in the code, it felt more obvious what they were. @f-f, do you want to give this a look, especially how I reworded SPEC.md?

File: SPEC.md [hunk 1]
```diff
@@ -199,10 +199,15 @@ All packages in the registry contain a `purs.json` manifest file in their root d
...
 - `dependencies`: dependencies of your package as key-value pairs where the keys are [`PackageName`](#packagename)s and values are [`Range`](#range)s; this is a required field, but if you have no dependencies you can provide an empty object.
-  - All dependencies you provide must exist in the registry, and the dependency ranges must be solvable (ie. it must be possible to produce a single version of each dependency that satisfies the provided version bounds, including any transitive dependencies).
+
+Note:
+
+- Globs you provide at the `includeFiles` and `excludeFiles` keys must contain only `*`, `**`, `/`, `.`, `..`, and characters for Linux file paths. It is not possible to negate a glob (ie. the `!` character), and globs cannot represent a path out of the package source directory.
+- When packaging your project source, the registry will first "include" your `src` directory and always-included files such as your `purs.json` file. Then it will include files which match globs indicated by the `includeFiles` key. Finally, it will apply the excluding globs indicated by the `excludeFiles` key to the included files (always-included files cannot be excluded).
```
Comment (f-f) [inline SPEC.md]: We should link to the list of "always included" files if we have it anywhere

File: app/src/App/API.purs [hunk 2]
```diff
@@ -899,33 +901,51 @@ formatPursuitResolutions { resolutions, dependenciesDir } =
...
+    Nothing -> pure []
+    Just nonEmptyGlobs -> do
+      let globs = map NonEmptyString.toString $ NonEmptyArray.toArray nonEmptyGlobs
+      { succeeded, failed } <- FastGlob.match source globs
+      let succeededAndNotExcluded = succeeded Array.\\ userExcludeFiles
 
       unless (Array.null failed) do
         Except.throw $ String.joinWith " "
-          [ "Some paths matched by globs in the 'files' key are outside your package directory."
+          [ "Some paths matched by globs in the 'include' key are outside your package directory."
```
Comment (f-f) [inline app/src/App/API.purs]:
```suggestion
          [ "Some paths matched by globs in the 'includeFiles' key are outside your package directory."
```

File: app/src/App/API.purs [hunk 3]
```diff
@@ -899,33 +901,51 @@ formatPursuitResolutions { resolutions, dependenciesDir } =
...
 
-      case NonEmptyArray.fromArray (Array.filter (Regex.test Internal.Path.pursFileExtensionRegex) succeeded) of
+      case NonEmptyArray.fromArray (Array.filter (Regex.test Internal.Path.pursFileExtensionRegex) succeededAndNotExcluded) of
         Nothing -> pure unit
         Just matches -> do
           let fullPaths = map (\path -> Path.concat [ source, path ]) matches
           Operation.Validation.validatePursModules fullPaths >>= case _ of
             Left formattedError ->
-              Except.throw $ "Some PureScript modules listed in the 'files' section of your manifest contain malformed or disallowed module names." <> formattedError
+              Except.throw $ "Some PureScript modules listed in the 'include' section of your manifest contain malformed or disallowed module names." <> formattedError
```
Comment (f-f) [inline app/src/App/API.purs]:
```suggestion
              Except.throw $ "Some PureScript modules listed in the 'includeFiles' section of your manifest contain malformed or disallowed module names." <> formattedError
```

File: SPEC.md (hunk 1 above)
Comment (thomashoneyman) [reply SPEC.md]: We do, and I've added links to both the always-included and always-ignored files.

**Review by thomashoneyman — APPROVED** [review]

**Review by f-f — APPROVED** [review]

**Pull request merged**